pkg/apis/extensions/types.go

```diff
@@ -691,3 +691,143 @@ const (
 	LabelSelectorOpDoesNotExist LabelSelectorOperator = "DoesNotExist"
 )
+
+// PodSecurityPolicy governs the ability to make requests that affect the SecurityContext
+// that will be applied to a pod and container.
+type PodSecurityPolicy struct {
+	unversioned.TypeMeta `json:",inline"`
+	api.ObjectMeta       `json:"metadata,omitempty"`
+
+	// Spec defines the policy enforced.
+	Spec PodSecurityPolicySpec `json:"spec,omitempty"`
+}
+
+// PodSecurityPolicySpec defines the policy enforced.
+type PodSecurityPolicySpec struct {
+	// Privileged determines if a pod can request to be run as privileged.
+	Privileged bool `json:"privileged,omitempty"`
+	// Capabilities is a list of capabilities that can be added.
+	Capabilities []api.Capability `json:"capabilities,omitempty"`
+	// Volumes is a white list of allowed volume plugins.  Empty indicates that all plugins
+	// may be used.
+	Volumes []FSType `json:"volumes,omitempty"`
+	// HostNetwork determines if the policy allows the use of HostNetwork in the pod spec.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
+	// HostPorts determines which host port ranges are allowed to be exposed.
+	HostPorts []HostPortRange `json:"hostPorts,omitempty"`
+	// HostPID determines if the policy allows the use of HostPID in the pod spec.
+	HostPID bool `json:"hostPID,omitempty"`
+	// HostIPC determines if the policy allows the use of HostIPC in the pod spec.
+	HostIPC bool `json:"hostIPC,omitempty"`
+	// SELinuxContext is the strategy that will dictate the allowable labels that may be set.
+	SELinuxContext SELinuxContextStrategyOptions `json:"seLinuxContext,omitempty"`
+	// RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
+	RunAsUser RunAsUserStrategyOptions `json:"runAsUser,omitempty"`
+}
+
+// HostPortRange defines a range of host ports that will be enabled by a policy
+// for pods to use.  It requires both the start and end to be defined.
+type HostPortRange struct {
+	// Start is the beginning of the port range which will be allowed.
+	Start int `json:"start"`
+	// End is the end of the port range which will be allowed.
+	End int `json:"end"`
+}
+
+// FSType gives strong typing to different file systems that are used by volumes.
+type FSType string
+
+var (
+	HostPath              FSType = "hostPath"
+	EmptyDir              FSType = "emptyDir"
+	GCEPersistentDisk     FSType = "gcePersistentDisk"
+	AWSElasticBlockStore  FSType = "awsElasticBlockStore"
+	GitRepo               FSType = "gitRepo"
+	Secret                FSType = "secret"
+	NFS                   FSType = "nfs"
+	ISCSI                 FSType = "iscsi"
+	Glusterfs             FSType = "glusterfs"
+	PersistentVolumeClaim FSType = "persistentVolumeClaim"
+	RBD                   FSType = "rbd"
+	Cinder                FSType = "cinder"
+	CephFS                FSType = "cephFS"
+	DownwardAPI           FSType = "downwardAPI"
+	FC                    FSType = "fc"
+)
```

Comment:
Until such a time as these volume types are removed from the core API for
pods and pvs, they ultimately need representation here.  So I agree it's
not very pluggable, but that's not a problem created by this PR or one that
should be resolved here.

One consideration is for custom volume plugins is whether the declared type
of the custom volume should be a string map declared here.  Custom volumes
have just as much potential for root compromise as any other type.

On Fri, Jan 8, 2016 at 1:18 PM, James DeFelice <notifications@github.com>
wrote:

> In pkg/apis/extensions/types.go
> <https://github.com/kubernetes/kubernetes/pull/7893#discussion_r49218945>:
>
> > +	HostPath              FSType = "hostPath"
> > +	EmptyDir              FSType = "emptyDir"
> > +	GCEPersistentDisk     FSType = "gcePersistentDisk"
> > +	AWSElasticBlockStore  FSType = "awsElasticBlockStore"
> > +	GitRepo               FSType = "gitRepo"
> > +	Secret                FSType = "secret"
> > +	NFS                   FSType = "nfs"
> > +	ISCSI                 FSType = "iscsi"
> > +	Glusterfs             FSType = "glusterfs"
> > +	PersistentVolumeClaim FSType = "persistentVolumeClaim"
> > +	RBD                   FSType = "rbd"
> > +	Cinder                FSType = "cinder"
> > +	CephFS                FSType = "cephFS"
> > +	DownwardAPI           FSType = "downwardAPI"
> > +	FC                    FSType = "fc"
> > +)
>
> this seems pretty inflexible, tying volume types directly into the API,
> and seems to contradict the idea of pluggable things - it's not very
> pluggable if it requires API changes.
> … <#506423031_>
> On Fri, Jan 8, 2016 at 11:12 AM, Eric Tune <notifications@github.com>
> wrote: In pkg/apis/extensions/types.go <#7893 (comment)
> <https://github.com/kubernetes/kubernetes/pull/7893#discussion_r49204717>>:
> > + HostPath FSType = "hostPath" > + EmptyDir FSType = "emptyDir" > +
> GCEPersistentDisk FSType = "gcePersistentDisk" > + AWSElasticBlockStore
> FSType = "awsElasticBlockStore" > + GitRepo FSType = "gitRepo" > + Secret
> FSType = "secret" > + NFS FSType = "nfs" > + ISCSI FSType = "iscsi" > +
> Glusterfs FSType = "glusterfs" > + PersistentVolumeClaim FSType =
> "persistentVolumeClaim" > + RBD FSType = "rbd" > + Cinder FSType = "cinder"
> > + CephFS FSType = "cephFS" > + DownwardAPI FSType = "downwardAPI" > + FC
> FSType = "fc" > +) @thockin <https://github.com/thockin> <
> https://github.com/thockin> @saad-ali <https://github.com/saad-ali> <
> https://github.com/saad-ali> did you see that the list of volume types is
> being enumerated here, and do you reluctantly accept it as necessary? —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/7893/files#r49204717>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/7893/files#r49218945>.
>